`Logger` proposal with `.d.ts` File

glenn2223 opened the pull request:
A proposal for a JS `Logging` mechanism plus a `.d.ts` file for shipping with the npm package

Refs: #2979, sass/dart-sass#13

glenn2223 committed:
Create JavaScript API Document

glenn2223 committed:
Update to API - more to come

glenn2223 committed:
Update Javascript API.md

glenn2223 committed:
Update Javascript API.md

glenn2223 committed:
Update Javascript API.md

glenn2223 committed:
Adding typescript declaration file

For JS API with intended `Logger` functionality

glenn2223 committed:
Update index.d.ts

glenn2223 committed:
Merge branch 'sass:master' into master

glenn2223 committed:
JS API

Create a JS API for current functionality plus a `Logger` proposal

glenn2223 committed:
Delete Javascript API.md

nex3 committed:
Add a TypeScript formatter and reformat the TS API

nex3 committed:
Merge branch 'master' of github.com:glenn2223/sass into glenn2223-master

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+  | {file: string}
+  | {contents: string}
+  | Error
+  | null;
```
`Error` isn't a valid synchronous return type for an `Importer` function (they should just throw instead), and `null` is only a valid return type for importers that are going to call their `done` callback (and even then only if they're passed to `render()`, not `renderSync()`). I think we should more clearly encode the distinction between synchronous and asynchronous importers:

```ts
export type Importer = (
  url: string,
  prev: string
) => {file: string} | {contents: string};

export type AsyncImporter = Importer | (
  url: string,
  prev: string,
  done: (result: {file: string} | {contents: string} | Error)
) => void;
```

Then you can make `renderSync()` take `Importer`s and `render()` take `AsyncImporter`s.

glenn2223 replied:
I gave the option for `render` and `renderSync` to support both async and sync calls. From what you're saying here, it can't so I'll update

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+  done: (data: ImporterReturnType) => void
+) => ImporterReturnType | void;
+
+export interface LogData {
```
Add some interface-level documentation as well.

Remember that this isn't just normal API documentation; this is also a specification for how implementation behave. Multiple implementations will need to support this API, and it's important that they do so as consistently as possible. There should be no room for ambiguity in how a given chunk of JS + Sass will work.

(This is mostly important for the logger-related interfaces, because that's a new API that needs to be implemented from scratch. The existing APIs that you're defining types here for would *ideally* be fully specified as well, but it's a lot less important because they're generally already implemented and because they'll be supplanted by #3056 before too long anyway.)

glenn2223 replied:
Added some documentation to each line - ready for review

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+  /**
+   * The message as-is in the file/data
+   */
+  message: string;
```
This should probably be passed as a positional parameter, to indicate that it's the most important part of each call.

glenn2223 replied:
Done

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+  /**
+   * A set of information related to the location of the error
+   */
+  source: SourceSpan;
```
This should be optional for `warn`. There are some warnings that don't have corresponding spans (such as a warning about how `render()` itself is being called).

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+  /**
+   * Is it a deprecation message
+   */
+  deprecation: boolean;
```
This field doesn't make sense for the `debug` function. There are enough differences between the two signatures that you should probably just make them both take inline options, rather than trying to share a single type that's passed to both.

glenn2223 replied:
Separated them into `DebugData` and `WarnData`

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+
+export interface Logger {
+  /**
+   * A function that is called each time `@warn` is encountered
```
This is not a comprehensive specification. `warn()` may also be called for warnings generated by the implementation itself, and in fact the implementation *must* pass all warnings to `warn()` if it's specified.

Which gets to another point: what should the implementation do if one of these functions is *not* specified and it needs to emit a warning or a `@debug` message?

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+   *
+   * @default null
+   */
+  data?: string;
```
Although each of these is individually optional, at least one of them and optionally both *must* be passed by the caller. You can see [here](https://github.com/sass/embedded-host-node/blob/661fa06465e8cae1ecc518e11dc6dae24dccb573/lib/src/node-sass/render.ts#L13-L44) how @awjin represents this for the Node embedded host.

glenn2223 replied:
Thanks for the sample - every day is a learning day 😁

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+   * A function that is called each time `@warn` is encountered
+   * @param data All data relative to the `@warn` line
+   */
+  warn?(data: LogData): void;
```
`warn()` should also take an optional stack trace.

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+  debug?(data: LogData): void;
+}
+
+export interface Options {
```
You'll probably need to split this into `Options` and `AsyncOptions` as well, which both inherit from `SharedOptions`.

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+  /**
+   * Path to a file to compile.
+   *
+   * @default null
```
It's not obvious how this `@default` tag maps onto the behavior that's required for an implementation.

glenn2223 replied:
Sorry, I don't follow

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+   */
+  functions?: {
+    [key: string]: (...args: types.SassType[]) => types.SassType | void;
+  };
```
Like importers, functions can be either synchronous or asynchronous. If they're asynchronous, the last argument is a `done()` callback.

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+  file: string;
+}
+
+export interface SourceLocation {
```
Like the `Logger` classes, these source interfaces should be thoroughly specified.

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+export function renderSync(options: Options): Result;
+
+export namespace types {
+  abstract class SassType {}
```
Dart Sass doesn't export any object with this name, and I don't *think* Node Sass does either.

glenn2223 replied:
This way just an easy way to add it to the `functions` parameter - it's not actually exported. 

The alternative is `types.Null | types.Number | types.String | ....`

How would you like to proceed?

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+    NULL: Null;
+  }
+
+  const Null: Null;
```
Because you're specifying the `Null` interface separately from the corresponding constant, I'm not sure if TypeScript will understand that `types.Null.NULL` is an instance of `types.Null`. You may need to just make this a `class` without a (visible) constructor.

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+
+  const Null: Null;
+
+  class Number implements SassType {
```
The linter is complaining that these are unused. Do you need to export them?

nex3 commented on index.d.ts:
```diff
@@ -0,0 +1,373 @@
+    setA(value: number): void;
+  }
+
+  class List<T extends SassType = SassType> implements SassType {
```
Sass doesn't consider its lists or maps to be typed, so I don't think this API should either.

glenn2223 replied:
So they should just be `List<unknown>` and `Map<unknown>`, is that right?

nex3 reviewed (changes requested):
Thanks for taking the time to write out the entire API!

glenn2223 committed:
Update index.d.ts

Commenting/documenting everything
Dropped unnecessary exports
Added exports to types
Various other changes - highlighted by @nex3

glenn2223 committed:
Update index.d.ts

Resolved https://github.com/sass/sass/pull/3067#discussion_r638380217

glenn2223 committed:
Match sass/master

glenn2223 committed:
Update index.d.ts

Fix https://github.com/sass/sass/pull/3067#discussion_r638377327
